format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/SplitMultilineIterator.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.delimited.input;
+
+import com.google.common.collect.AbstractIterator;
+import org.apache.hadoop.mapreduce.RecordReader;
+
+import java.io.IOException;
+
+/**
+ * Iterator that provides the splits in the delimited string based on the delimiter. The delimiter
+ * should not contain any quotes. The spliterator will behave like this: 1. if there is no quote, it
+ * will behave same as {@link String#split(String)} 2. if there are quotes in the string, the method
+ * will find pairs of quotes, content within each pair of quotes will not get split even if there
+ * is delimiter in that. For example, if string is a."b.c"."d.e.f" and delimiter is '.', it will get
+ * split into [a, b.c, d.e.f]. if string is "val1.val2", then it will not get split since the '.'
+ * is within a pair of quotes. If data is spread across multiple lines and quote is started within first line, it
+ * will iterate over next lines until closing quote is there.If the delimited string contains odd number of quotes,
+ * which mean the quotes are not closed, an exception will be thrown. The quote within the value will always be trimmed.
+ */
+public class SplitMultilineIterator extends AbstractIterator<String> {
```

Comment:
this class looks like it has a good amount of duplicate code with the SplitQuotesIterator. There should be a way to enhance the current iterator, or create a common base class instead of creating a completely separate class.

The concern here is if we add more options in the future (ex: ability to configure the quote character, allow escaping quotes, etc.), the two classes can easily get out of sync.

Reply:
added to a single class.